framework/ui/view/launchprovider.cpp

```diff
@@ -137,7 +163,66 @@ RetVal<Val> QmlLaunchProvider::toRetVal(const QVariant& jsrv) const
     return rv;
 }
 
-void QmlLaunchProvider::onOpen(QString pageType)
+RetVal<LaunchProvider::OpenData> LaunchProvider::openWidget(const UriQuery &q)
+{
+    RetVal<OpenData> result;
+
+    QString uri = QString::fromStdString(q.uri().toString());
+
+    ContainerMeta container = uriRegister()->container(uri);
+    int widgetMetaTypeId = container.widgetMetaTypeId;
+
+    QString objectId = QString::number(widgetMetaTypeId); // unque?
+
+    void *widgetClassPtr = QMetaType::create(widgetMetaTypeId);
+    QDialog* dialog = static_cast<QDialog*>(widgetClassPtr);
+
+    if (!dialog) {
+        result.ret = make_ret(Ret::Code::UnknownError);
+        return result;
+    }
+
+    fillData(dialog, q);
+
+    QVariantMap status;
+    connect(dialog, &QDialog::finished, [=, &status, &objectId](int finishStatus){
+        status["errcode"] = static_cast<int>(Ret::Code::Ok);
+        status["value"] = finishStatus;
+
+        onPopupClose(objectId, status);
+    });
+
+    onOpen(PAGE_TYPE_WIDGET);
+
+    dialog->exec();
```

Comment:
Here you need to either TODO or immediately do this:

```
if (sync) {
    dialog->exec();
} else {
    dialog->show();
}
```
And add dialog delete to its close signal (remember to use deleteLater)

Reply:
Added asynchronous opening.
As for the dialog delete, it is destroyed using QMetaType::destroy.